src/module.cpp

```diff
@@ -533,11 +533,129 @@ void Module::AddGlobalVariable(const std::string &name, const Type *type, Expr *
+                        std::vector<llvm::APFloat> fpVal;
+                        int i = 0;
+                        bool canInitVaryingConst = true;
+                        for (const Expr *e : exprList->exprs) {
+                            const ConstExpr *ce = llvm::dyn_cast<ConstExpr>(e);
+                            if (!ce || ce->Count() != 1) {
+                                canInitVaryingConst = false;
+                                break;
+                            }
+                            const Type *t = ce->GetType();
+                            if (!t || !t->IsUniformType() || !t->IsAtomicType()) {
+                                canInitVaryingConst = false;
+                                break;
+                            }
+                            switch (basicType) {
+                            case AtomicType::TYPE_BOOL:
+                                ce->GetValues(&vals_bool[i++]);
+                                break;
+                            case AtomicType::TYPE_INT8:
+                                ce->GetValues(&vals_int8[i++]);
+                                break;
+                            case AtomicType::TYPE_UINT8:
+                                ce->GetValues(&vals_uint8[i++]);
+                                break;
+                            case AtomicType::TYPE_INT16:
+                                ce->GetValues(&vals_int16[i++]);
+                                break;
+                            case AtomicType::TYPE_UINT16:
+                                ce->GetValues(&vals_uint16[i++]);
+                                break;
+                            case AtomicType::TYPE_INT32:
+                                ce->GetValues(&vals_int32[i++]);
+                                break;
+                            case AtomicType::TYPE_UINT32:
+                                ce->GetValues(&vals_uint32[i++]);
+                                break;
+                            case AtomicType::TYPE_INT64:
+                                ce->GetValues(&vals_int64[i++]);
+                                break;
+                            case AtomicType::TYPE_UINT64:
+                                ce->GetValues(&vals_uint64[i++]);
+                                break;
+                            case AtomicType::TYPE_FLOAT16:
+                            case AtomicType::TYPE_FLOAT:
+                            case AtomicType::TYPE_DOUBLE: {
+                                std::vector<llvm::APFloat> ce_vals;
+                                ce->GetValues(ce_vals);
+                                fpVal.push_back(ce_vals[0]);
+                                break;
+                            }
+                            default:
+                                // Unsupported types.
+                                break;
+                            }
+                        }
+                        if (canInitVaryingConst) {
```

Comment:
What would happen if expr list is too short for the varying? Can it happen?
What would happen if the types in expr list are different types? Would the convert handled correctly?

Reply:
> What would happen if expr list is too short for the varying?

The other elements are initialized with defaults (zeros, false).

> Can it happen?

Yes, see `tests/lit-tests/2738_partial_init.ispc`

> What would happen if the types in expr list are different types?

Const values would be converted in `GetValues` methods of `ConstExpr`.

> Would the convert handled correctly?

See tests: 
- tests/lit-tests/2738_mixed_types.ispc
- tests/lit-tests/2738_cast_types.ispc

Reply:
> > What would happen if expr list is too short for the varying?
> 
> The other elements are initialized with defaults (zeros, false).

I changed that in the last revision to match the behavior of other initializations, i.e., if too short but not 1 then emit error otherwise initialize the whole vector with the same value:

Reply:
> 

> I changed that in the last revision to match the behavior of other initializations, i.e., if too short but not 1 then emit error otherwise initialize the whole vector with the same value

This definitely needs to be documented in the language reference.

Reply:
I have documented this.